### DISCUSSION TITLE
Discussion: How can we create a "V Shell" to replace bash, zsh, fish, etc. for daily use?

### DISCUSSION THREAD

**Discussion by ylluminate** [vlang/v#11681] [Ideas]
It would be nice to use V everywhere - even in every day shell (Terminal / cmd prompt) scenarios.  

There has been some questions, thoughts, and expressions of desire from time to time about how we might go about creating a V language shell that has enough expression and terseness to allow users to use V directly in a shell so as to not feel the need to use BASH, ZSH or other shells.  This is a bit tricky presently since it appears that the syntax is not directly translatable for more interactive and constant use (yes, it's fine for building shell scripts as already exists).

While outdated / defunct, `rush` (https://github.com/adamwiggins/rush) was an interesting such similarly goal'ed project.

Anyone with thoughts or ideas in this regard is more than welcome to dump them here for discussion and consideration!

**Comment by ylluminate** [comment]
Interesting read, but perhaps not entirely pertinent to this particular topic: https://presstige.io/p/Using-Go-instead-of-bash-for-scripts-6b51885c1f6940aeb40476000d0eb0fc

**Comment by THX1188** [comment]
That would be great, but there were some tries to make a high-level-program-language oriented shells, but they never made it, for example the csh. To be a Shell you have to be in some kind easy-going to be _shellish_ that way. Even v is tiny and small it's declarative syntax and abscence of globals makes it hard to be shellish. A broad amount of Shell-Users are not programmers. 
But I would like it, as it guy.

**Comment by elimisteve** [comment]
> Interesting read, but perhaps not entirely pertinent to this particular topic: https://presstige.io/p/Using-Go-instead-of-bash-for-scripts-6b51885c1f6940aeb40476000d0eb0fc

Or, use both Bash and Go programs -- together :slightly_smiling_face: https://github.com/progrium/go-basher

**Comment by elimisteve** [comment]
A few years ago, Brian Fox (creator of Bash) said "I wish that thing I made in the '80s would just die a quiet death" and that someone would make a shell better suited to the current programming environment.

He said Bash was optimized for dealing with text and files and directories, because that's what mattered then. But he wishes we had **"Nesh", a "network shell", that was optimized to make it really easy to script interactions between servers, containers, and VMs**.

_Really_ interesting! And a good example of how our tooling has completely failed to keep pace with the environment in which we work.

**Comment by elimisteve** [comment]
Seems like a "Nesh" could let us do Ansible-type stuff in a few keystrokes...

Separate idea: what if we could access V commands at a shell, and indicate this by beginning those functions with `v.`, as in `v.net.http.get Github.com`.

**Comment by elimisteve** [comment]
**A next-generation shell should make it easy to pipe together web service APIs as easily as we pipe together local commands today.** That's part of the idea behind https://webpipes.org , too, though that idea isn't shell-specific, it's more about writing much less code.

But it'd be way better to have that functionality built into a shell! :tada: 

> $ Twitter @elimisteve .tweets --latest .url | Discord v #programming message "Here is the URL to my latest tweet: $stdout"

---

> What sort of info must go into the Webpipes spec? We should be able to use machine-readable schemas for much of what Webpipe clients need to parse.
>
> There's the rules that specify exactly how outputs may be transformed in order to become a new input, like if `text, err` is returned from one endpoint then we'd maybe check the error, return that error if it's non-empty, then pass along text to the next pipe in the chain...

Not that `text, err` is in a Go context. For V that would mean a function returning a `?string`, printing the error to stderr if the optional's `error` is not `none` then exiting, or piping along.

Another idea I mentioned a year ago to the main Webpipes dev:

> ...I just realized that we could have some sick look-ahead logic that either spits out `{"text": "the answer"}` or `"the answer"` depending upon what the next pipe function expects!!

**Comment by elimisteve** [comment]
For data querying and wrangling we could have built-in support for GROQ, which is brilliantly simple and powerful -- a near-impossible combination: https://groq.dev/ .

If you have JSON like

```json
[
  {
    "userId": 1,
    "id": 3,
    "title": "fugiat veniam minus",
    "completed": false
  },
  {
    "userId": 1,
    "id": 4,
    "title": "et porro tempora",
    "completed": true
  },
  ...
]
```

and you want an array of the titles of the completed tasks, GROQ lets you do

    *[completed == true && userId == 2][].title

That's it!!!

**Comment by elimisteve** [comment]
If such a shell parsed Ansible inventory files and `~/.ssh/.config`, and also gave us a place to name remote services (kind of like git remotes) and credentials, and if those remote services used OpenAPI and were thus machine-readable, wow... Then a nesh could bootstrap its capabilities instantly.

Throw in some smarts into the shell itself, like with look-ahead conventions (like knowing whether to pass text or JSON from one command's output to become the next command's input), plus really succinct syntax and smart defaults... Sounds compelling.

**Comment by elimisteve** [comment]
"Take this photo from my Google Drive, pass it through a photo filter's API, then post the resulting photo to my Instagram" should be trivial and not require tons of manually glue code!

**Comment by lygstate** [comment]
get v shell in better shape and redesign

**Comment by shadowninja55** [comment]
here's a similar idea in python: https://xon.sh/

**Comment by ylluminate** [comment]
Really nice observation / callout there @shadowninja55. I think this could provide a really nice model. @medvednikov what do you think?: 
![example](https://p153.p0.n0.cdn.getcloudapp.com/items/NQuoPXzj/9799e7c5-913f-47d8-ad98-b6fa43eff355.jpg?v=c6c62e6350cd0dc108a77dd313101ff6)
